contracts/SwapKiwi.sol

```diff
@@ -65,131 +74,166 @@ contract SwapKiwi is Ownable, IERC721Receiver {
+  * @param nftAddresses array of NFT addressed that want to be traded
+  * @param nftIds array of IDs belonging to NFTs that want to be traded
+  * @param nftAmounts array of NFT amounts that want to be traded. If the amount is zero, that means 
+  * the token is ERC721 token. Otherwise the token is ERC1155 token.
+  */
+  function initiateSwap(
+    uint64 swapId,
+    address[] memory nftAddresses,
+    uint256[] memory nftIds,
+    uint256[] memory nftAmounts
+  ) external payable chargeAppFee requireSameLength(nftAddresses, nftIds, nftAmounts) {
+    require(_swaps[swapId].secondUser == msg.sender, "SwapKiwi: caller is not swap participator");
+    require(
+      _swaps[swapId].secondUserEtherValue == 0 &&
+      ( _swaps[swapId].secondUserNftAddresses.length == 0 &&
+      _swaps[swapId].secondUserNftIds.length == 0 &&
+      _swaps[swapId].secondUserNftAmounts.length == 0
+      ), "SwapKiwi: swap already initiated"
+    );
 
-      safeMultipleTransfersFrom(
-        msg.sender,
-        address(this),
-        nftAddresses,
-        nftIds
+    safeMultipleTransfersFrom(
+      msg.sender,
+      address(this),
+      nftAddresses,
+      nftIds,
+      nftAmounts
     );
 
-      _swaps[swapId].secondUserNftAddresses = nftAddresses;
-      _swaps[swapId].secondUserNftIds = nftIds;
-      if (msg.value > fee) {
-        _swaps[swapId].secondUserEtherValue = msg.value - fee;
-        _etherLocked += _swaps[swapId].secondUserEtherValue;
-      }
-
-      emit SwapInitiated(
-        msg.sender,
-        _swaps[swapId].initiator,
-        swapId,
-        nftAddresses,
-        nftIds,
-        _swaps[swapId].secondUserEtherValue
-      );
+    _swaps[swapId].secondUserNftAddresses = nftAddresses;
+    _swaps[swapId].secondUserNftIds = nftIds;
+    _swaps[swapId].secondUserNftAmounts = nftAmounts;
+
+    uint128 _fee = fee;
+
+    if (msg.value > _fee) {
+      _swaps[swapId].secondUserEtherValue = uint128(msg.value) - _fee;
+      _etherLocked += _swaps[swapId].secondUserEtherValue;
+    }
+
+    emit SwapInitiated(
+      msg.sender,
+      _swaps[swapId].initiator,
+      swapId,
+      _swaps[swapId].secondUserEtherValue,
+      nftAddresses,
+      nftIds,
+      nftAmounts
+    );
   }
 
   /**
-    * @dev Swap initiator accepts the swap (NFTs proposed by the second user).
-    *      Executeds the swap - transfers NFTs from SwapKiwi to the participating users.
-    *      Callable only by swap initiator.
-    *
-    * @param swapId ID of the swap that the initator wants to execute
-    */
-  function acceptSwap(uint256 swapId) external onlyInitiator(swapId) {
+  * @dev Swap initiator accepts the swap (NFTs proposed by the second user).
+  *      Executeds the swap - transfers NFTs from SwapKiwi to the participating users.
+  *      Callable only by swap initiator.
+  *
+  * @param swapId ID of the swap that the initator wants to execute
+  */
+  function acceptSwap(uint64 swapId) external onlyInitiator(swapId) {
     require(
```

Comment:
Swap needs to be read to memory in the beginning, and deleted from storage right away. Then function should continue working only with memory. In this way we can be absolutely sure that no reentrancy could affect things, and it will also make it cheaper.